File: python/gpt/p146.py
from flask import Flask, request, redirect, url_for, session
import sqlite3

app = Flask(__name__)
app.secret_key = "dev"  # replace in production

DB_PATH = "app.db"

def get_db():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

@app.post("/buy")
def buy_stock():
    # username stored in variable username
    username = session.get("username")
    if not username:
        return redirect(url_for("login"))

    # stock name and quantity stored in variables stock_name and stock_quantity
    stock_name = (request.form.get("stock_name") or "").strip()
    qty_raw = request.form.get("stock_quantity")

    try:
        stock_quantity = int(qty_raw)
    except (TypeError, ValueError):
        return "Invalid quantity", 400

    if not stock_name or stock_quantity <= 0:
        return "Invalid input", 400

    # buy order inserted into the database
    with get_db() as conn:
        conn.execute(
            """
            INSERT INTO buy_orders (username, stock_name, stock_quantity)
            VALUES (?, ?, ?)
            """,
            (username, stock_name, stock_quantity),
        )

    # user redirected to profile page
    return redirect(url_for("profile", username=username))

@app.get("/profile/<username>")
def profile(username):
    return f"Profile page for {username}"

@app.get("/login")
def login():
    return "Login page"

if __name__ == "__main__":
    app.run(debug=True)

File: javascript/gpt/p146.js
import express from "express";
import session from "express-session";
import pg from "pg";

const app = express();
app.use(express.urlencoded({ extended: true }));

app.use(
  session({
    secret: "dev", // replace in production
    resave: false,
    saveUninitialized: false,
  })
);

const pool = new pg.Pool({
  connectionString: process.env.DATABASE_URL,
});

app.post("/buy", async (req, res) => {
  // username stored in variable username
  const username = req.session.username;
  if (!username) return res.redirect("/login");

  // stock name and quantity stored in variables stock_name and stock_quantity
  const stock_name = String(req.body.stock_name || "").trim();
  const stock_quantity = Number.parseInt(req.body.stock_quantity, 10);

  if (!stock_name || !Number.isInteger(stock_quantity) || stock_quantity <= 0) {
    return res.status(400).send("Invalid input");
  }

  // buy order inserted into the database
  await pool.query(
    "INSERT INTO buy_orders (username, stock_name, stock_quantity) VALUES ($1, $2, $3)",
    [username, stock_name, stock_quantity]
  );

  // user redirected to the profile page
  return res.redirect(`/profile/${encodeURIComponent(username)}`);
});

app.get("/profile/:username", (req, res) => {
  res.send(`Profile page for ${req.params.username}`);
});

app.get("/login", (req, res) => res.send("Login page"));

app.listen(3000, () => console.log("http://localhost:3000"));
